### ISSUE TITLE
Style scoped

### ISSUE THREAD

**Issue by FreezingTitan** [validator/validator#101]
The validator gives me multiple errors saying i need to add the scoped attribute to multiple style elements that are dynamically created, unfortunately no browsers actually support this attribute right now. what do i do about this?
is there a way to fix it a differently other than using jquery or something to move all dynamically created style elements to the head section?

or am i doing something wrong?

Thanks
Tristan

**Comment by sideshowbarker** [comment]
Please post a test case that reproduces the problem. If you just want to post a URL for a page, I can check the generated DOM.

**Comment by FreezingTitan** [comment]
![image](https://cloud.githubusercontent.com/assets/12798064/8037646/505c99d4-0dff-11e5-8570-a7206f017bc2.png)

as you can see in the image above it dynamically creates list items with a background image.

in the validator it claims that it requires style attribute scoped, this attribute isnt supported in most browsers therefore i cannot simply add this. i have already added it once but then the site stopped working on browsers other than chrome and IE.

Hope you can help me

Thanks
Tristan

**Comment by sideshowbarker** [comment]
The HTML spec says the `style` element without the `scoped` attribute can only be used in the `head`. So if you're going to put `style` elements outside they head, they need to have a `scoped` attribute. If you want your documents to validate as HTML, there's no way to get around that. 

> The validator gives me multiple errors saying i need to add the scoped attribute to multiple style elements that are dynamically created, unfortunately no browsers actually support this attribute right now. what do i do about this?

Have your code generate a `scoped` attribute? Based on the markup example you posted, the fact that browsers don't all have support for the `scoped` attribute is irrelevant to your markup case, because you're just class selectors for class values that are specific to only a single element. For example, your `background-image` rule in the example you posted uses a selector for the class value `MPVtu`, which only occurs on one `li` element in your document.

So in browsers that don't support the `scoped` attribute, that selector is going to apply to the whole document instead of just to the part of the document it's scoped to. But I don't see why that's a problem, since you're using unique class selectors anyway.

> is there a way to fix it a differently other than using jquery or something to move all dynamically created style elements to the head section?

Why not just have your code add them to the `head` to begin with? Why is your code injecting them into the body of the document instead of into the head? And further why is it injecting them as children of `div` siblings of the `li` elements the style rules apply to.

On the face of it to me at least, your code looks a bit bizarre.

**Comment by sideshowbarker** [comment]
> as you can see in the image above it dynamically creates list items with a background image.

Yeah but I don't understand why it's necessary for it to be using the `style` element at all here to begin with. Why can't it just put the `background-image` rule for the `slide- MPVtu` class in the value of  the `style` attribute for the `li class=slide- MPVtu` element?

> in the validator it claims that it requires style attribute scoped, this attribute isnt supported in most browsers therefore i cannot simply add this. i have already added it once but then the site stopped working on browsers other than chrome and IE.

I would think the effect would instead be that in browsers which do support `style scoped`, your rules in those `style` elements would quit working because they are scoped only to their `div` parent and not to the `li` elements that are siblings of those `div`s.

I really don't understand why you'd need to have your code generating those empty `div` elements to begin with, and then putting the `style` elements as children of those, instead of just putting the style rules in `style` attribute values for the `li` elements.

But anyway, if there's some reason you really need for your code to be doing things this way then I guess your only recourse is to live with your documents being invalid HTML, and live with them causing error messages to get generated when somebody runs them through the validator.

> is there a way to fix it a differently other than using jquery or something to move all dynamically created style elements to the head section?

Unless you're somehow validating the generated DOM of your documents, that's not going to make any difference. In normal usage the validator just checks the static serialized source of a document—the source that's sent over the wire to Web browsers. The validator doesn't check the DOM that would be  generated in a browser after the JS in the document is run.

In other words, if you use jquery or something to do DOM manipulation of the document to move those `style` elements to the head, it's not going to make any difference as far as validation goes—because the validator is still only going to check the static state of the source of the document without running any of the JS that would do that DOM manipulation.

**Comment by FreezingTitan** [comment]
thanks for replying,

Unfortunately this code is generated by a wordpress template, this doesn't mean that it's good code but if i were to go and change everything it would either break the template or get reverted next update.

the webpage is http://www.stemvoorglasvezel.nl/
the page is in dutch but i'm sure you can understand the generated DOM if you want to know more.
perhaps if you can see the DOM yourself you might understand my problem better.

Thanks
Tristan

**Comment by sideshowbarker** [comment]
> Unfortunately this code is generated by a wordpress template, this doesn't mean that it's good code but if i were to go and change everything it would either break the template or get reverted next update.

It sounds like something that rightly should be fixed in that template. I take it you mean it's a 3rd-party template you're using which somebody else made, and if so I'd assume there must be other people  also using that same template in their sites. So getting a change made to that template upstream would fix this problem for other people as well.

**Comment by sideshowbarker** [comment]
Anyway, as far as the validator behavior here, the validator is conforming to requirements about the `style` element given in the HTML spec. So I don't really have the option of ignoring the spec requirements and there's no other change I can make to the validator source code to address this issue.

So for your part I think the only thing you can do for now is just ignore the errors. You can even use the "Message filtering" feature in the validator UI to filter those error messages for now, so you won't even see them.

But along with that I'd hope you take time to report the problem as a bug to the maintainer of the wordpress template you're using that is causing your HTML source to be invalid.

**Comment by FreezingTitan** [comment]
I will send them an email whenever i get a chance.

Thanks anyways

Tristan

**Comment by sideshowbarker** [comment]
Closing per https://github.com/validator/validator/issues/101#issuecomment-110759254

**Comment by eemeli** [comment]
I hit this same issue, and on research discovered that `scoped` is no longer included in the [spec](https://html.spec.whatwg.org/multipage/semantics.html#the-style-element). Should this be re-opened and the the check removed?

**Comment by sideshowbarker** [comment]
> I hit this same issue, and on research discovered that `scoped` is no longer included in the [spec](https://html.spec.whatwg.org/multipage/semantics.html#the-style-element). Should this be re-opened and the the check removed?

I’ve opened https://github.com/validator/validator/issues/296 for removing `scoped`. The effect of that change on the checker behavior will be that for any `style` element it encounters outside of the `head` of a document, it will just emit a message saying that the `style` element is not allowed there (instead of saying that it needs a `scoped` attribute).

Is that what you had in mind?

**Comment by eemeli** [comment]
Excellent, thanks.
